external/fv3fit/fv3fit/emulation/data/config.py

```diff
@@ -0,0 +1,245 @@
+import dacite
+import dataclasses
+import logging
+import xarray
+import yaml
+import numpy as np
+import tensorflow as tf
+from toolz.functoolz import compose_left
+from typing import Any, Callable, Dict, Mapping, Sequence, Tuple, Union
+
+from . import transforms
+
+
+logger = logging.getLogger(__name__)
+
+
+class SliceLoader(yaml.SafeLoader):
+    """
+    Extended safe yaml loader to interpret slices from a sequence.
+
+    Example
+    -------
+
+    The vertical subselection transform yaml::
+
+        name: maybe_subselect
+        args:
+          - specific_humidity: !!python/slice [15, null]
+            tendency_of_sphum: !!python/slice [15, null]
+
+    Loading the yaml file::
+
+        with open("yaml_file_with_slices.yaml", "r") as f:
+            yaml.load(f, Loader=SliceLoader)
+
+    """
+
+    def construct_python_slice(self, node):
+        return slice(*self.construct_sequence(node))
+
+
+SliceLoader.add_constructor(
+    "tag:yaml.org,2002:python/slice", SliceLoader.construct_python_slice
+)
+
+
+Dataset = Union[xarray.Dataset, Mapping[str, np.ndarray], Mapping[str, tf.Tensor]]
+XyTensors = Tuple[Tuple[tf.Tensor], Tuple[tf.Tensor]]
+
+
+@dataclasses.dataclass
+class _TransformConfigItem:
+    """
+    Specification of a transform function and any
+    args and/or kwargs necessary to curry it into a
+    function taking a single argument.
+
+    Args
+    ----
+        name: Curry-able function name in the transforms module
+        args: Arguments to provide into curried function such that
+            the resulting call signature is func(dataset).  Note:
+            This can be specified as a mapping but it will be converted
+            to a list in the order supplied.
+        kwargs: Keyword arguments to provide to the curried function.
+
+    Example
+    -------
+    Example yaml for grouped inputs/outputs transform::
+
+        name: group_inputs_outputs
+        args:
+          - ["field1", "field2"]
+          - ["field3"]
+
+    Equivalent yaml for grouped inputs/outputs transform::
+
+        name: group_inputs_outputs
+        args:
+          input_variables: ["field1", "field2"]
+          output_variables: ["field3"]
+    """
+
+    name: str
+    args: Union[Sequence[Any], Mapping[str, Any]] = dataclasses.field(
+        default_factory=list
+    )
+    kwargs: Mapping[str, Any] = dataclasses.field(default_factory=dict)
+
+    def __post_init__(self):
+        if isinstance(self.args, Mapping):
+            self.args = list(self.args.values())
```

Comment:
See comment about removal.  I originally had this in there because I was only using mappings (easier to read from yaml) for argument input into the functions.  I think I ever bring config specified custom pipelines, I'll make sure it can just use all keyword arguments instead.